Make errors when building with MAGMA support (and fix)

When running with `--with-magma` and properly set include and libs, I get this error when running `make`:
```
In file included from dsyr2k.c:8:
In file included from ./f2c.h:22:
../utilities/_hypre_utilities.h:100:4: error: unknown type name 'magma_queue_t'
  100 |    magma_queue_t          magma_queue;
      |    ^
```

This is because in `handle.h:83`, `magma_queue_t` is used prior to including `<magma_v2.h>`:
```
// ...
#if defined(HYPRE_USING_MAGMA)
   magma_queue_t          magma_queue;
#endif
} hypre_Handle;
```

Adding these lines to `header.h` fixed the issue for me:
```diff
diff --git a/src/utilities/handle.h b/src/utilities/handle.h
index 44ce6d382..ec456545b 100644
--- a/src/utilities/handle.h
+++ b/src/utilities/handle.h
@@ -25,6 +25,9 @@
 #endif /* UMPIRE_VERSION_MAJOR >= 2022 */
 #define HYPRE_UMPIRE_POOL_NAME_MAX_LEN 1024
 #endif /* defined(HYPRE_USING_UMPIRE) */
+#if defined(HYPRE_USING_MAGMA)
+struct magma_queue_t;
+#endif
 
 struct hypre_DeviceData;
 typedef struct hypre_DeviceData hypre_DeviceData;
```

Just as well, you could probably just move `magma.h` up in the include order in `src/utilities/headers`:
```diff
diff --git a/src/utilities/headers b/src/utilities/headers
index 841f787ff..4a615cd5e 100755
--- a/src/utilities/headers
+++ b/src/utilities/headers
@@ -37,11 +37,11 @@ extern "C" {
 # Structures and prototypes
 #===========================================================================
 
+cat magma.h                    >> $INTERNAL_HEADER
 cat handle.h                   >> $INTERNAL_HEADER
 cat state.h                    >> $INTERNAL_HEADER
 cat general.h                  >> $INTERNAL_HEADER
 cat base.h                     >> $INTERNAL_HEADER
-cat magma.h                    >> $INTERNAL_HEADER
 cat matrix_stats.h             >> $INTERNAL_HEADER
 cat printf.h                   >> $INTERNAL_HEADER
 cat error.h                    >> $INTERNAL_HEADER
```

--- a/src/config/cmake/HYPRE_CMakeUtilities.cmake
+++ b/src/config/cmake/HYPRE_CMakeUtilities.cmake
@@ -22,6 +22,47 @@ function(set_hypre_option category name description default_value)
   endif()
 endfunction()
 
+# Scope interface -std flags to the appropriate languages so C files do not
+# inherit C++ standard flags from TPLs (e.g., pkg-config Cflags like -std=c++14)
+function(scope_language_standard_flags TARGET_NAME)
+  if(NOT TARGET ${TARGET_NAME})
+    return()
+  endif()
+
+  get_target_property(_opts ${TARGET_NAME} INTERFACE_COMPILE_OPTIONS)
+  if(NOT _opts)
+    return()
+  endif()
+
+  set(_new_opts)
+  foreach(_o IN LISTS _opts)
+    # Preserve existing generator expressions untouched
+    if("${_o}" MATCHES "^\$<")
+      list(APPEND _new_opts "${_o}")
+      continue()
+    endif()
+
+    set(_is_std_flag FALSE)
+    if("${_o}" MATCHES "^-std=")
+      set(_is_std_flag TRUE)
+    endif()
+
+    if(_is_std_flag)
+      # Classify by presence of '++' (C++) vs C
+      string(FIND "${_o}" "++" _has_plusplus)
+      if(_has_plusplus GREATER -1)
+        list(APPEND _new_opts "$<$<OR:$<COMPILE_LANGUAGE:CXX>,$<COMPILE_LANGUAGE:CUDA>,$<COMPILE_LANGUAGE:HIP>>:${_o}>")
+      else()
+        list(APPEND _new_opts "$<$<COMPILE_LANGUAGE:C>:${_o}>")
+      endif()
+    else()
+      list(APPEND _new_opts "${_o}")
+    endif()
+  endforeach()
+
+  set_target_properties(${TARGET_NAME} PROPERTIES INTERFACE_COMPILE_OPTIONS "${_new_opts}")
+endfunction()
+
 # Function to set internal hypre build options
 function(set_internal_hypre_option var_prefix var_name)
   if(HYPRE_ENABLE_${var_name})
@@ -394,6 +435,8 @@ function(setup_tpl LIBNAME)
           else()
             target_link_libraries(${PROJECT_NAME} PUBLIC ${LIBNAME}::${LIBNAME})
           endif()
+          # Ensure imported standard flags are scoped to valid languages
+          scope_language_standard_flags(${LIBNAME}::${LIBNAME})
           message(STATUS "Found ${LIBNAME} target: ${LIBNAME}::${LIBNAME}")
         elseif(TARGET ${LIBNAME})
           if(${LIBNAME_UPPER} STREQUAL "UMPIRE")
@@ -406,6 +449,8 @@ function(setup_tpl LIBNAME)
           else()
             target_link_libraries(${PROJECT_NAME} PUBLIC ${LIBNAME})
           endif()
+          # Ensure imported standard flags are scoped to valid languages
+          scope_language_standard_flags(${LIBNAME})
           message(STATUS "Found ${LIBNAME} target: ${LIBNAME}")
         else()
           message(FATAL_ERROR "${LIBNAME} target not found. Please check your ${LIBNAME} installation")
@@ -470,6 +515,8 @@ function(setup_tpl LIBNAME)
               # Prefer linking the generated IMPORTED target when available
               if(TARGET PkgConfig::PC_${LIBNAME_UPPER})
                 target_link_libraries(${PROJECT_NAME} PUBLIC PkgConfig::PC_${LIBNAME_UPPER})
+                # Prevent C++ standard flags from leaking into C compiles
+                scope_language_standard_flags(PkgConfig::PC_${LIBNAME_UPPER})
               else()
                 target_link_libraries(${PROJECT_NAME} PUBLIC ${PC_${LIBNAME_UPPER}_LINK_LIBRARIES})
               endif()

--- a/src/utilities/_hypre_utilities.h
+++ b/src/utilities/_hypre_utilities.h
@@ -25,6 +25,79 @@ extern "C" {
  * SPDX-License-Identifier: (Apache-2.0 OR MIT)
  ******************************************************************************/
 
+#ifndef HYPRE_MAGMA_HEADER
+#define HYPRE_MAGMA_HEADER
+
+#include "HYPRE_config.h"
+
+#if defined(HYPRE_USING_MAGMA)
+
+#include "error.h"
+
+#ifdef __cplusplus
+extern "C++"
+{
+#endif
+
+#if !defined(MAGMA_GLOBAL)
+#if !defined(ADD_)
+#define ADD_
+#endif
+#endif
+#include <magma_v2.h>
+
+#ifdef __cplusplus
+}
+#endif
+
+/*--------------------------------------------------------------------------
+ * Wrappers to MAGMA functions according to hypre's precision
+ *--------------------------------------------------------------------------*/
+
+#if defined(HYPRE_COMPLEX) || defined(HYPRE_LONG_DOUBLE)
+#error "MAGMA interface does not support (yet) HYPRE_COMPLEX and HYPRE_LONG_DOUBLE"
+
+#elif defined(HYPRE_SINGLE)
+#define hypre_magma_getrf_gpu              magma_sgetrf_gpu
+#define hypre_magma_getrf_nat              magma_sgetrf_native
+#define hypre_magma_getrs_gpu              magma_sgetrs_gpu
+#define hypre_magma_getri_gpu              magma_sgetri_gpu
+#define hypre_magma_getri_nb               magma_get_dgetri_nb
+#define hypre_magma_gemv                   magma_sgemv
+
+#else /* Double precision */
+#define hypre_magma_getrf_gpu              magma_dgetrf_gpu
+#define hypre_magma_getrf_nat              magma_dgetrf_native
+#define hypre_magma_getrs_gpu              magma_dgetrs_gpu
+#define hypre_magma_getri_gpu              magma_dgetri_gpu
+#define hypre_magma_getri_nb               magma_get_sgetri_nb
+#define hypre_magma_gemv                   magma_dgemv
+
+#endif
+
+/*--------------------------------------------------------------------------
+ * General wrapper call to MAGMA functions
+ *--------------------------------------------------------------------------*/
+
+#define HYPRE_MAGMA_CALL(call) do {                   \
+   magma_int_t err = call;                            \
+   if (MAGMA_SUCCESS != err) {                        \
+      printf("MAGMA ERROR (code = %d) at %s:%d\n",    \
+            err, __FILE__, __LINE__);                 \
+      hypre_assert(0);                                \
+   } } while(0)
+
+#define HYPRE_MAGMA_VCALL(call) call
+
+#endif /* HYPRE_USING_MAGMA */
+#endif /* HYPRE_MAGMA_HEADER */
+/******************************************************************************
+ * Copyright (c) 1998 Lawrence Livermore National Security, LLC and other
+ * HYPRE Project Developers. See the top-level COPYRIGHT file for details.
+ *
+ * SPDX-License-Identifier: (Apache-2.0 OR MIT)
+ ******************************************************************************/
+
 /******************************************************************************
  *
  * General structures and values
@@ -422,77 +495,6 @@ typedef struct
 #define hypre_SolverDestroy(data)     ((data) -> destroy)
 
 #endif /* HYPRE_BASE_HEADER */
-/******************************************************************************
- * Copyright (c) 1998 Lawrence Livermore National Security, LLC and other
- * HYPRE Project Developers. See the top-level COPYRIGHT file for details.
- *
- * SPDX-License-Identifier: (Apache-2.0 OR MIT)
- ******************************************************************************/
-
-#ifndef HYPRE_MAGMA_HEADER
-#define HYPRE_MAGMA_HEADER
-
-#include "HYPRE_config.h"
-
-#if defined(HYPRE_USING_MAGMA)
-
-#include "error.h"
-
-#ifdef __cplusplus
-extern "C++"
-{
-#endif
-
-#if !defined(MAGMA_GLOBAL)
-#define ADD_
-#endif
-#include <magma_v2.h>
-
-#ifdef __cplusplus
-}
-#endif
-
-/*--------------------------------------------------------------------------
- * Wrappers to MAGMA functions according to hypre's precision
- *--------------------------------------------------------------------------*/
-
-#if defined(HYPRE_COMPLEX) || defined(HYPRE_LONG_DOUBLE)
-#error "MAGMA interface does not support (yet) HYPRE_COMPLEX and HYPRE_LONG_DOUBLE"
-
-#elif defined(HYPRE_SINGLE)
-#define hypre_magma_getrf_gpu              magma_sgetrf_gpu
-#define hypre_magma_getrf_nat              magma_sgetrf_native
-#define hypre_magma_getrs_gpu              magma_sgetrs_gpu
-#define hypre_magma_getri_gpu              magma_sgetri_gpu
-#define hypre_magma_getri_nb               magma_get_dgetri_nb
-#define hypre_magma_gemv                   magma_sgemv
-
-#else /* Double precision */
-#define hypre_magma_getrf_gpu              magma_dgetrf_gpu
-#define hypre_magma_getrf_nat              magma_dgetrf_native
-#define hypre_magma_getrs_gpu              magma_dgetrs_gpu
-#define hypre_magma_getri_gpu              magma_dgetri_gpu
-#define hypre_magma_getri_nb               magma_get_sgetri_nb
-#define hypre_magma_gemv                   magma_dgemv
-
-#endif
-
-/*--------------------------------------------------------------------------
- * General wrapper call to MAGMA functions
- *--------------------------------------------------------------------------*/
-
-#define HYPRE_MAGMA_CALL(call) do {                   \
-   magma_int_t err = call;                            \
-   if (MAGMA_SUCCESS != err) {                        \
-      printf("MAGMA ERROR (code = %d) at %s:%d\n",    \
-            err, __FILE__, __LINE__);                 \
-      hypre_assert(0);                                \
-   } } while(0)
-
-#define HYPRE_MAGMA_VCALL(call) call
-
-#endif /* HYPRE_USING_MAGMA */
-#endif /* HYPRE_MAGMA_HEADER */
 /******************************************************************************
  * Copyright (c) 1998 Lawrence Livermore National Security, LLC and other
  * HYPRE Project Developers. See the top-level COPYRIGHT file for details.

--- a/src/utilities/headers
+++ b/src/utilities/headers
@@ -37,11 +37,11 @@ extern "C" {
 # Structures and prototypes
 #===========================================================================
 
+cat magma.h                    >> $INTERNAL_HEADER
 cat handle.h                   >> $INTERNAL_HEADER
 cat state.h                    >> $INTERNAL_HEADER
 cat general.h                  >> $INTERNAL_HEADER
 cat base.h                     >> $INTERNAL_HEADER
-cat magma.h                    >> $INTERNAL_HEADER
 cat matrix_stats.h             >> $INTERNAL_HEADER
 cat printf.h                   >> $INTERNAL_HEADER
 cat error.h                    >> $INTERNAL_HEADER

--- a/src/utilities/magma.h
+++ b/src/utilities/magma.h
@@ -20,7 +20,9 @@ extern "C++"
 #endif
 
 #if !defined(MAGMA_GLOBAL)
+#if !defined(ADD_)
 #define ADD_
+#endif
 #endif
 #include <magma_v2.h>
 

